### ISSUE TITLE
mbox from google takeout-not working?

### ISSUE THREAD

**Issue by nazgul6** [coandco/gtalk_export#2]
Hi,

I've download the .mbox of my chats using google takeout.

I'm using Terminal on OSX. 

This is the result I get:

Processing mbox file at /XXX/Takeout/Mail/Chat.mbox
Traceback (most recent call last):
  File "/Users/XXX/gtalk_export-master/gtalk_export.py", line 139, in <module>
    parse_mbox(args.mbox_location, args.name, args.email, args.timestamp_format)
  File “/XXX/gtalk_export.py", line 49, in parse_mbox
    name = re.sub("Chat with ", "", message['subject'])
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 155, in sub
    return _compile(pattern, flags).sub(repl, string, count)
TypeError: expected string or buffer

(I've edited out some of the file paths above and replaced them with XXX)

Any ideas as to what I could be doing wrong / how to fix this? 

Thanks.

**Comment by coandco** [comment]
Hmm.  The error suggests that the message['subject'] object that's being passed to the regex substitution is something other than a string, which theoretically shouldn't happen if the mbox file has been prepared correctly.  Either way, try replacing the following line:

```
        name = re.sub("Chat with ", "", message['subject'])
```

With this:

```
        try:
            name = re.sub("Chat with ", "", message['subject'])
        except TypeError:
            print("TypeError, subject is %s" % message['subject'])
            continue
```

That should skip over any messages that have a bad subject line, rather than just crashing and stopping the whole process.  It'll also output the bad subject lines to the console when you run it.

**Comment by nazgul6** [comment]
Thanks!

I did just as said and I now get 'TypeError, subject is None' about millions of times in Terminal, followed by:

Traceback (most recent call last):
  File "/Users/XXX/gtalk_export.py", line 143, in <module>
    parse_mbox(args.mbox_location, args.name, args.email, args.timestamp_format)
  File "/Users/XXX/gtalk_export.py", line 63, in parse_mbox
    timestamp = time.strftime(timestamp_format, parsedate(rawtimestr))
TypeError: argument must be 9-item sequence, not None

There are files output in the folder I was CD'in in.

**Comment by coandco** [comment]
Okay, so I'm looking at this a bit more in depth now, something sticks out to me.  Where exactly did you get your .mbox file?  gtalk_export takes (up to) two files: 1) A .json file obtained from Google Takeout, and 2) a .mbox file generated by connecting to the GMail account in question with an IMAP client and downloading only the "Chats" folder.  You appear to be attempting to use some sort of a .mbox file taken directly from Takeout.

**Comment by nazgul6** [comment]
Thanks for the reply. Yes, as mentioned, I used the Google Takeout for both the Chat.mbox (taken out as a specific, single label from the gmail - I figured this would be a faster way than using a mail client to build the Chat.mbox) and the Hangouts (for the .json).

**Comment by coandco** [comment]
_nods_ I suspect that's the source of the issues.  I'll take a gander at what sort of formatting the Takeouts version of the .mbox produces on my own account to see how it differs.

**Comment by nazgul6** [comment]
Thanks - I'm curious about the findings.

**Comment by coandco** [comment]
Takeouts is taking a long time to export my Chats mbox, so it may be a while.

**Comment by nazgul6** [comment]
No worries. Mine was 33mb - took minutes:)

**Comment by coandco** [comment]
Okay, I've finally gotten around to looking at the .mbox files generated by Takeouts, and it looks like Google does some additional stripping that makes at least some of the messages useless for archival purposes..  Specifically, the gtalk_export .mbox parser looks for one of two things: a) a multipart message, which indicates a pre-changeover full-conversation XML message, or b) a single-part message, which indicates a post-changeover singleton-style message.  In the case of a single-part message, it needs to see both a "To" and a "From" field to determine who the conversation is between.

The chat .mbox file I got from Takeouts appears to have a bunch of messages that lack subject lines, date fields, and "To" fields.  These are the messages that were causing the errors you were seeing in the first post of this issue.  There's nothing I can do to process those, as they don't have necessary information.  I'm not sure whether using the Thunderbird-export method will grab these in a viable way.

That said, the Takeouts .mbox file did still have some messages in the format I'm expecting.  That's why the script was able to process some chats when you made the try/except change.

tl;dr It looks like newer chats are butchered by the Takeouts export process.  It's probably worth trying the Thunderbird-export method to see if that results in a viable .mbox file.

**Comment by nazgul6** [comment]
Hi, Thanks for the testing done. The Thunderbird way worked for me in the end (after also trying Mail.app, whose .mbox eport didn't work either). 

The script did, however, split some of my conversations into multiple .txts, but I take it that is due to me having my gmail interface in 3 different languages over the years so gmail must labeled/subjected my chats in slightly different ways depending on the language. I guess I'll just manually cut and paste those .txt logs together.
